src/Npgsql/Replication/ReplicationConnection.cs

```diff
@@ -229,26 +221,25 @@ public async Task OpenAsync(CancellationToken cancellationToken = default)
         public ValueTask DisposeAsync()
         {
             using (NoSynchronizationContextScope.Enter())
-                return DisposeAsyncInternal();
-        }
-
-        async ValueTask DisposeAsyncInternal()
-        {
-            var currentState =
-                (ReplicationConnectionState)Interlocked.Exchange(ref _state, (int)ReplicationConnectionState.Disposed);
-            if (currentState == ReplicationConnectionState.Disposed)
-                return;
+                return DisposeAsyncCore();
 
-            if (State == ReplicationConnectionState.Streaming)
+            async ValueTask DisposeAsyncCore()
             {
-                Connector.CancelRequest();
-                await _replicationCompletion;
-            }
+                if (_isDisposed)
```

Comment:
Are you sure this is enough?
Can't there be a race condition if two threads enter `DisposeAsync()`?
The previous disposed checking code looked a bit safer to me.

Reply:
You're right that the previous code was safer for concurrent disposing, but in general we aren't safe in Npgsql against concurrency, so I didn't put in the effort to protect against it here specifically; that's precisely what we should focus on in vNext, but for Npgsql as a whole. This is again the case with many other standard types, e.g. SemaphoreSlim.

If you feel there's a special reason to guard against two threads in DisposeAsync we can have a volatile flag for that - but it would be an exceptional protection for a specific place.

Reply:
> If you feel there's a special reason to guard against two threads in DisposeAsync

No I don't see why we have special needs here compared to other places within Npgsql. I was just wondering if I have some kind of fundamental misunderstanding about how things work because you replaced the previously thread-safe code with non-thread-safe code and it would be pretty easy to make it thread safe again.
I have no problem with doing it systematically for all of Npgsql at some point.